### PR TITLE
`merkledb` --  fix `findNextKey`

### PR DESCRIPTION
**Pull request by danlaine**
## Why this should be merged

Fixes a bug.

## How this works

It's possible for the `endProof` passed into `findNextKey` to be length 0. In this case we  panic on `if !proofKeyPath.HasPrefix(receivedProofNodes[len(receivedProofNodes)-1].KeyPath) {`

This PR fixes the bug by handling this case.

Also improves comment on `findNextKey`.

## How this was tested

New UT, also separate merkledb testing tool

### CODE REVIEW

**Review by dboehm-avalabs — APPROVED** [review]

File: x/sync/syncmanager.go [hunk 1]
```diff
@@ -344,14 +344,29 @@ func (m *StateSyncManager) getAndApplyRangeProof(ctx context.Context, workItem *
 	m.completeWorkItem(ctx, workItem, largestHandledKey, rootID, proof.EndProof)
 }
 
-// findNextKey attempts to find the first key larger than the lastReceivedKey that is different in the local merkle vs the merkle that is being synced.
-// Returns the first key with a difference in the range (lastReceivedKey, rangeEnd), or nil if no difference was found in the range
+// findNextKey returns the start of the key range that should be fetched next.
+// Returns nil if there are no more keys to fetch  up to [rangeEnd].
+//
+// If the last proof received contained at least one key-value pair, then
+// [lastKeyReceived] is the greatest key in the key-value pairs received.
```
Comment (darioush) [inline x/sync/syncmanager.go:351]:
```suggestion
// [lastReceivedKey] is the greatest key in the key-value pairs received.
```

File: x/sync/syncmanager.go [hunk 2]
```diff
@@ -344,14 +344,29 @@ func (m *StateSyncManager) getAndApplyRangeProof(ctx context.Context, workItem *
...
 	ctx context.Context,
 	lastReceivedKey []byte,
 	rangeEnd []byte,
-	receivedProofNodes []merkledb.ProofNode,
+	endProof []merkledb.ProofNode,
 ) ([]byte, error) {
+	if len(endProof) == 0 {
+		// We try to find the next key to fetch by looking at the end proof.
+		// If the end proof is empty, we have no information to use.
+		return lastReceivedKey, nil
```
Comment (darioush) [inline x/sync/syncmanager.go:367]: should we append `byte(0)` to `lastReceivedKey` in this case?
In which case would the proof not at least include the root?

File: x/sync/syncmanager.go [hunk 3]
```diff
@@ -344,14 +344,29 @@ func (m *StateSyncManager) getAndApplyRangeProof(ctx context.Context, workItem *
...
+// Returns nil if there are no more keys to fetch  up to [rangeEnd].
+//
+// If the last proof received contained at least one key-value pair, then
+// [lastKeyReceived] is the greatest key in the key-value pairs received.
+// Otherwise it's the end of the range for the last proof received.
+//
+// [rangeEnd] is the end of the range that we want to fetch.
+//
+// [endProof] is the end proof of the last proof received.
+// Namely it's a proof for [lastKeyReceived].
```
Comment (darioush) [inline x/sync/syncmanager.go:357]: It's also possible it's an exclusion proof if the next key is outside of rangeEnd

File: x/sync/syncmanager.go (hunk 2 above)
Comment (danlaine) [reply x/sync/syncmanager.go:367]: > should we append byte(0) to lastReceivedKey in this case?

I don't think this works. Suppose the `lastReceivedKey` is `[1,2,3]`. There could be keys that begin `[1,2,4]` and we would skip those if we returned `[1,2,3,0]`

Comment (danlaine) [reply x/sync/syncmanager.go]: > should we append byte(0) to lastReceivedKey in this case?
Done.

> In which case would the proof not at least include the root?

See [here](https://github.com/ava-labs/avalanchego/pull/1629/files#diff-dee93f9a6a8e07479e6553b5276c89123d5b7c18b233c262e6f3cd700fb6b2fcR255) for a more clear statement of the end proof invariants than in this PR's code

**Pull request merged**